connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java

```diff
@@ -81,13 +86,18 @@ public void startSampling() {
    */
   public void stopSampling() {
     if (mSamplingCounter.decrementAndGet() == 0) {
-      mHandler.sendEmptyMessage(SamplingHandler.MSG_STOP);
+      mHandler.removeMessages(SamplingHandler.MSG_START);
+      addSample();
+      /**
+       * Resets previously read byte count after recording a sample, so that
+       * we don't count bytes downloaded in between sampling sessions.
+       */
+      sPreviousBytes = -1;
     }
   }
 
   private class SamplingHandler extends Handler {
     static final int MSG_START = 1;
```

Comment:
FWIW, the better pattern here is to make this private and add an explicit method to control the messages that go in (`sendSend` in this case).  That way, the sendStart method can do the necessary logic around removing the existing message.  In general, you want to keep the exact message semantics internal/private to the handler and expose a structured Java API only.